### PR TITLE
ISO 8601 date format support

### PR DESCRIPTION
**Pull request by RobertTheProfessional**
ISO 8601 is an international standard for date and time data and is widely used in other languages such as Java and Javascript. Having a formatter in the core would help with deserialization to the Date Foundation struct.

### CODE REVIEW

**Comment by vzsg** [comment]
Did you test this on Linux?
Because I'm quite sure that `Calendar(identifier:)` explodes pretty hard.

**Comment by RobertTheProfessional** [comment]
You're absolutely correctly the `Calendar(identifier:)` line definitely blows up my test API quickly in Linux. After commenting out the line the Core package builds and my test API responds with the correct ISO8601 string. I have to double check as many test cases as possible to see what adverse affects could occur.

Appears they've acknowledged the [bug](https://bugs.swift.org/browse/SR-3828) and [merged](https://github.com/apple/swift-corelibs-foundation/pull/880) it into the corelibs-foundation

**Comment by RobertTheProfessional** [comment]
Fortunately, the Calendar defaults to the Gregorian Calendar which is the requirement for the ISO8601 standard. Therefore the explicit Calendar assignment isn't required. I'll update the pull request accordingly.

**Comment by LoganWright** [comment]
We have some patterns here that I think make discovery easier and more useful for date formatters https://github.com/vapor/node/blob/26b0547ac4bd5ce8f7ba0217371bdee6c8f1c886/Sources/Node/Convertibles/Date%2BConvertible.swift

This way, one can type `DateFormatter.` and get automatic completion for the type of formatter they might be looking for. I think I'd prefer to bring some of these date formatters into Core with this pattern as opposed to creating new types etc. for each potential formatter.

Any thoughts on that, and thanks for the PR proposal 👍

**Review by loganwright — COMMENTED** [review]
We have some patterns here that I think make discovery easier and more useful for date formatters https://github.com/vapor/node/blob/26b0547ac4bd5ce8f7ba0217371bdee6c8f1c886/Sources/Node/Convertibles/Date%2BConvertible.swift

This way, one can type DateFormatter. and get automatic completion for the type of formatter they might be looking for. I think I'd prefer to bring some of these date formatters into Core with this pattern as opposed to creating new types etc. for each potential formatter.

Any thoughts on that, and thanks for the PR proposal 👍

**Comment by RobertTheProfessional** [comment]
I didn't realize the formatter was already extended to support ISO 8601 in the vapor/node package. Considering the current state of the Core package allows for direct initialization of a Date object via a RFC 1123 string; initialization via an ISO 8601 would fit in well versus having an end-user extend the Date object themselves for initialization.

Personally, I believe having common and popular Date formats should be in the Core package, then niche formats such as (MySQL or PostgreSQL) can extend from their own packages.

 What's your take?

File: Tests/CoreTests/ISO8601Tests.swift [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+	static var allTests = [
+		("testBasic", testBasic)
+	]
+	
+	func testBasic() {
+		let date = Date(iso8601: "2016-07-01T04:00:00.000Z")
+		XCTAssertEqual(date?.iso8601, "2016-07-01T04:00:00.000Z")
+	}
+	
+	func testFail() {
```
Comment (kgn) [inline Tests/CoreTests/ISO8601Tests.swift:17]: `testFail` should be added to `allTests`

File: Sources/Core/ISO8601.swift [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import Foundation
+
+public struct ISO8601 {
+	public static let shared = ISO8601()
+	public let formatter: DateFormatter
+	
+	public init() {
+		let formatter = DateFormatter()
```
Comment (kgn) [inline Sources/Core/ISO8601.swift]: Why not just set `formatter`?

``` swift
		formatter = DateFormatter()
  		formatter.locale = Locale(identifier: "en_US_POSIX")
 		formatter.timeZone = TimeZone(secondsFromGMT: 0)
 		formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSXXXXX"
```

**Comment by kgn** [comment]
I'd vote for `ISO8601` in Core

**Comment by RobertTheProfessional** [comment]
Added the failing ISO 8601 test case to allTests! Any reason why the RFC 1123 test case doesn't have its failing test case in its allTests variable?

**Comment by kgn** [comment]
@RobertTheProfessional  I believe it's just an oversight. Are you using tabs in your PR? I believe Vapor is all 4 spaces for indentation.

**Comment by RobertTheProfessional** [comment]
Yes I was using tabs! Just converted to spaces!

**Review by tanner0101 — APPROVED** [review]
LGTM.

**Comment by hhanesand** [comment]
Should the date formatter be threadsafe?

**Comment by tanner0101** [comment]
@hhanesand IMO the user of this should have to add locks if they need them (given that it's quite a performance hit). However, it would be good to warn in the docblock if usage of a function is not thread safe.

**Comment by RobertTheProfessional** [comment]
@hhanesand Can I get some clarification on the context in which thread safety would be an issue? The formatter itself is immutable and simply facilitates the extension of the Date object to allow for ISO 8601 instantiation. There shouldn't be an instance in which thread safety is an issue unless the developer explicitly tries to share the memory space.

**Comment by vzsg** [comment]
> The formatter itself is immutable

That's an invalid assumption. DateFormatter is not thread-safe on Linux.

**Comment by RobertTheProfessional** [comment]
I stand corrected! DateFormatter is only guaranteed on macOS 10.9 and later in a 64-bit application.

**Comment by kgn** [comment]
Do the thread concerns apply to the current implementation of `RFC 1123` as well? This new file is copied from that implementation correct? https://github.com/vapor/core/blob/master/Sources/Core/RFC1123.swift

**Comment by vzsg** [comment]
Actually, yeah, it should also apply there.

**Comment by hhanesand** [comment]
The Node implementation uses locks : https://github.com/vapor/node/blob/master/Sources/Node/Convertibles/Date%2BConvertible.swift#L66

Maybe we could include a global per-thread lock free date formatter for use of all vapor packages here in Core?

**Comment by tanner0101** [comment]
We're moving to Codable in 3.0 & Swift 4. See https://developer.apple.com/documentation/foundation/jsonencoder/2895363-dateencodingstrategy

**Pull request closed without merging**